src/prefect_server/api/states.py

```diff
@@ -100,6 +123,101 @@ async def set_flow_run_state(
             return_exceptions=True,
         )
 
+    # --------------------------------------------------------
+    # Queueing runs if using concurency limits
+    # --------------------------------------------------------
+
+    # If the run is not labeled or we've already checked during,
+    # a previous recursive call, we avoid this code branch
+    # as it's manually injecting a delay (and it isn't concurrency limited)
+    if (
+        (state.is_running() or state.is_submitted())
+        and flow_run.labels
+        and not has_been_concurrency_checked
+    ):
+        # Flow Concurrency Limits
+        # If the run is already occupying a slot (Submitted -> Running)
+        # or is attempting to occupy a slot (X -> Submitted)
+        existing_state = state_schema.load(flow_run.serialized_state)
+
+        # Due to the Agent process submitting multiple runs from Scheduled
+        # to Submitted at once, we need to acquire a lock to both check &
+        # set the new state if there are concurrency limits in place to
+        # avoid race conditions between the time a check has occurred and
+        # when an update has occurred.
+        server_context = context.get_context()
+        flow_concurrency_lock = server_context.get(
+            "flow_concurrency_lock", flow_concurrency_lock
+        )
+
+        async with flow_concurrency_lock:
+
+            can_transition = (
+                await api.flow_concurrency_limits.try_take_flow_concurrency_slots(
+                    tenant_id=flow_run.tenant_id,
+                    limit_names=flow_run.labels,
+                    current_state=existing_state,
+                )
+            )
+            if can_transition:
+
+                # If the flow run can transition, we continue holding the lock and
+                # call this pipeline recursively to bypass the checks
+                return await api.states.set_flow_run_state(
+                    flow_run_id=flow_run_id,
+                    state=state,
+                    version=version,
+                    agent_id=agent_id,
+                    has_been_concurrency_checked=True,
+                )
+
+        if existing_state.is_scheduled() and state.is_submitted():
+            buffer_seconds = random.randint(15, 45)
+
+            state = Queued(
```

Comment:
This use of `Queued` deviates from how Cloud uses `Queued`. This scenario occurs when an Agent attempts to submit a run and there are no slots left. If we raise an error here, the run will be put into a `Failed` state by the Agent, which isn't what we want.

`buffer_seconds` was a guesstimate of a decent implementation. It would be nice to add a jitter in here to avoid agents getting the same batches of runs over and over though.